engines/lab/lab.cpp

```diff
@@ -0,0 +1,252 @@
+	//SearchMan.addSubDirectoryMatching(gameDataDir, "game");
+	//SearchMan.addSubDirectoryMatching(gameDataDir, "game/pict");
+	//SearchMan.addSubDirectoryMatching(gameDataDir, "game/spict");
+	//SearchMan.addSubDirectoryMatching(gameDataDir, "music");
+}
+
+LabEngine::~LabEngine() {
+	// Remove all of our debug levels here
+	DebugMan.clearAllDebugChannels();
+
+	freeMapData();
+	for (uint16 i = 1; i <= _manyRooms; i++)
+		_resource->freeViews(i);
+	delete[] _rooms;
+	delete[] _inventory;
+
+	delete _conditions;
+	delete _roomsFound;
+	delete _event;
+	delete _resource;
+	delete _music;
+	delete _anim;
+	delete _graphics;
+	delete _tilePuzzle;
+	delete _utils;
+	delete _console;
+	delete _journalBackImage;
+	// _screenImage->_imageData is always pointing to the current drawing buffer.
+	// It shouldn't be deleted there.
+	_screenImage->_imageData = nullptr;
+	delete _screenImage;
+}
+
+Common::Error LabEngine::run() {
+	if (getFeatures() & GF_LOWRES)
+		initGraphics(320, 200, false);
+	else
+		initGraphics(640, 480, true);
+
+	_event = new EventManager(this);
+	_resource = new Resource(this);
+	_music = new Music(this);
+	_graphics = new DisplayMan(this);
+	_anim = new Anim(this);
+	_tilePuzzle = new TilePuzzle(this);
+	_utils = new Utils(this);
+	_console = new Console(this);
+	_journalBackImage = new Image(this);
+	_screenImage = new Image(this);
+
+	if (getPlatform() == Common::kPlatformWindows) {
+		// Check if this is the Wyrmkeep trial
+		Common::File roomFile;
+		bool knownVersion = true;
+		bool roomFileOpened = roomFile.open("game/rooms/48");
+
+		if (!roomFileOpened)
+			knownVersion = false;
+		else if (roomFile.size() != 892)
+			knownVersion = false;
+		else {
+			roomFile.seek(352);
+			byte checkByte = roomFile.readByte();
+			if (checkByte == 0x00) {
+				// Full Windows version
+			} else if (checkByte == 0x80) {
+				// Wyrmkeep trial version
+				_extraGameFeatures = GF_WINDOWS_TRIAL;
+
+				GUI::MessageDialog trialMessage("This is a trial Windows version of the game. To play the full version, you will need to use the original interpreter and purchase a key from Wyrmkeep");
+				trialMessage.runModal();
+			} else {
+				knownVersion = false;
+			}
+
+			roomFile.close();
+
+			if (!knownVersion) {
+				warning("Unknown Windows version found, please report this version to the ScummVM team");
+				return Common::kNoGameDataFoundError;
+			}
+		}
+	}
+
+	go();
+
+	return Common::kNoError;
+}
+
+Common::String LabEngine::generateSaveFileName(uint slot) {
+	return Common::String::format("%s.%03u", _targetName.c_str(), slot);
+}
+
+void LabEngine::drawStaticMessage(byte index) {
+	_graphics->drawMessage(_resource->getStaticText((StaticText)index), false);
+}
+
+void LabEngine::changeVolume(int delta) {
+	warning("STUB: changeVolume()");
+}
+
+void LabEngine::waitTOF() {
+	g_system->copyRectToScreen(_graphics->_displayBuffer, _graphics->_screenWidth, 0, 0, _graphics->_screenWidth, _graphics->_screenHeight);
```

Comment:
`Engine` keeps a `_system` pointer for accessing `OSystem` functionality. Thus, there's no need to use the global object here.